pip/vcs/git.py

```diff
@@ -74,27 +74,34 @@ def export(self, location):
                 show_stdout=False, cwd=temp_dir.path
             )
 
-    def check_rev_options(self, rev, dest, rev_options):
-        """Check the revision options before checkout to compensate that tags
-        and branches may need origin/ as a prefix.
-        Returns the SHA1 of the branch or tag if found.
+    def get_ref_sha(self, dest, ref):
         """
-        revisions = self.get_short_refs(dest)
-
-        origin_rev = 'origin/%s' % rev
-        if origin_rev in revisions:
-            # remote branch
-            return [revisions[origin_rev]]
-        elif rev in revisions:
-            # a local tag or branch name
-            return [revisions[rev]]
-        else:
+        Return a commit hash for the reference if the given name is a ref.
+        Otherwise, return None.
+
+        Args:
+          dest: the repository directory.
+          ref: the ref name.
+        """
+        output = self.run_command(['show-ref', '--hash', ref], cwd=dest,
+                                  show_stdout=False, on_returncode='ignore')
+        # The git-show-ref command can return multiple references, so we
+        # return only the first.
+        for sha in output.strip().splitlines():
+            return sha
```

Comment:
Thanks. I also found a simpler case involving a single branch in the "update" case. A local branch will shadow any updates to a remote after the initial install because "refs/heads" precedes "refs/remotes" alphabetically.

I do still think the current code can be made much simpler, easier to understand, and better / more easily tested. It is currently quite difficult to comprehend and to know what's necessary versus an artifact. I will update.